### PR TITLE
test comprenhensions (list, dict, gen exp) + tests

### PR DESCRIPTION
**Pull request by filipsch**
@vincentvankrunkelsven  can you have a quick look, this solves the issues #60 and #61

### CODE REVIEW

File: pythonwhat/parsing.py [hunk 1]
```diff
@@ -455,6 +449,86 @@ def visit_Lambda(self, node):
...
+                "list_comp": node,
+                "key": node.key,
+                "value": node.value,
+                "target": target,
+                "target_vars": get_target_vars(target),
+                "iter": node.generators[0].iter,
+                "ifs": node.generators[0].ifs
+            })
+
+class GeneratorExpParser(CompParser):
```
Comment (vincentvankrunkelsven) [inline pythonwhat/parsing.py]: Seems exactly the same as `ListCompParser`.

File: pythonwhat/parsing.py [hunk 2]
```diff
@@ -455,6 +449,86 @@ def visit_Lambda(self, node):
...
+        self.comps.append({
+                "list_comp": node,
+                "body": node.elt,
+                "target": target,
+                "target_vars": get_target_vars(target),
+                "iter": node.generators[0].iter,
+                "ifs": node.generators[0].ifs
+            })
+
+def get_target_vars(target):
```
Comment (vincentvankrunkelsven) [inline pythonwhat/parsing.py]: I'd make it a method of some superclass of the classes that use it.

File: pythonwhat/test_comp.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+              ifs=ifs,
+              insufficient_ifs_msg=insufficient_ifs_msg,
+              expand_message=expand_message,
+              student_comp_list=student_list_comps,
+              solution_comp_list=solution_list_comps,
+              state=state,
+              rep=rep,
+              comp_type="list")
+
+def test_generator_exp(index=1,
```
Comment (vincentvankrunkelsven) [inline pythonwhat/test_comp.py]: `test_generator_exp` and `test_list_comp` seem exactly the same except for `comp_type` (if the comment above about the parsers being the same was valid).

File: pythonwhat/test_comp.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+    state = State.active_state
+    rep = Reporter.active_reporter
+    rep.set_tag("fun", "test_list_comp")
+
+    state.extract_list_comps()
+    student_list_comps = state.student_list_comps
+    solution_list_comps = state.solution_list_comps
+
+    test_comp(index=index,
```
Comment (vincentvankrunkelsven) [inline pythonwhat/test_comp.py]: I'd programmatically add the arguments of the calling function, this seems error prone.

**Comment by vincentvankrunkelsven** [comment]
Some small remarks but generally looks good

File: pythonwhat/parsing.py (hunk 1 above)
Comment (filipsch) [inline pythonwhat/parsing.py]: It is not; `node.key`, `node.value` instead of `node.elt`.

File: pythonwhat/test_comp.py (hunk 3 above)
Comment (filipsch) [inline pythonwhat/test_comp.py]: yeah but you need different automatic messages... I already put it in `test_comp`, but yeah..I want to make `test_generator_exp` and `test_list_comp` available to the SCT writer..

File: pythonwhat/test_comp.py (hunk 4 above)
Comment (filipsch) [inline pythonwhat/test_comp.py]: how do you mean?

File: pythonwhat/parsing.py (hunk 2 above)
Comment (filipsch) [inline pythonwhat/parsing.py]: Can you be more specific?

File: pythonwhat/parsing.py (hunk 1 above)
Comment (vincentvankrunkelsven) [inline pythonwhat/parsing.py]: I was referring to `GeneratorExpParser`

File: pythonwhat/parsing.py (hunk 2 above)
Comment (vincentvankrunkelsven) [inline pythonwhat/parsing.py]: Well in this case for example, you're using `get_target_vars()` in `ForParser` and `CompParser`, so the closest common superclass is `Parser`. So in this case I'd make `get_target_vars(target)` a static method in parser. This works as well, it's just less OO.

**Pull request merged**